Add relref linting to lint script

sean1588 opened the pull request:
Part of: https://github.com/pulumi/registry/issues/1928
fixes: https://github.com/pulumi/registry/issues/1980

This PR just adds relref linting to the lint script and turns off currently failing linting rules. This lint script is currently not turned on in the pipeline. I also added a fix funtionality that can be run with make lint-fix. I ran it to fix up the relrefs that were failing here.

sean1588 commented:
here is a list of all the current failures:
```
Errors:

/home/work/pulumi/registry/themes/default/content/registry/packages/aquasec/installation-configuration.md:
Line 18: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/aviatrix/_index.md:
Line 10: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below].

/home/work/pulumi/registry/themes/default/content/registry/packages/aviatrix/installation-configuration.md:
Line 25: Lists should be surrounded by blank lines.
Line 30: Lists should be surrounded by blank lines.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws/_index.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 11: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 13: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 16: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 117: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 123: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 124: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws/installation-configuration.md:
Line 8: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 182: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 189: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-apigateway/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-iam/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-miniflux/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-native/_index.md:
Line 13: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 18: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 20: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 28: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 29: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 30: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-native/installation-configuration.md:
Line 8: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 138: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 142: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 149: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-quickstart-aurora-postgres/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-quickstart-redshift/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-quickstart-vpc/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/aws-s3-replicated-bucket/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/awsx/_index.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 14: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 16: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/awsx/installation-configuration.md:
Line 8: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure/_index.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 15: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 47: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure/installation-configuration.md:
Line 154: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-native/_index.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 11: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 107: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 113: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 117: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 121: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 123: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 124: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 125: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 126: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 132: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-native/from-classic.md:
Line 6: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-native/installation-configuration.md:
Line 165: Trailing spaces [Expected: 0 or 2; Actual: 38].
Line 160: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-native/version-guide.md:
Line 82: Multiple consecutive blank lines [Expected: 1; Actual: 2].
Line 160: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-quickstart-acr-geo-replication/installation-configuration.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line File Header: Page title exceeds 60 characters.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-static-website/_index.md:
Line 66: Multiple consecutive blank lines [Expected: 1; Actual: 2].
Line 71: Files should end with a single newline character.

/home/work/pulumi/registry/themes/default/content/registry/packages/azure-static-website/installation-configuration.md:
Line 38: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below].
Line 39: Fenced code blocks should be surrounded by blank lines.

/home/work/pulumi/registry/themes/default/content/registry/packages/azuread/installation-configuration.md:
Line 140: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/buildkite/installation-configuration.md:
Line 56: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/confluent/_index.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/confluent/installation-configuration.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/doppler/installation-configuration.md:
Line 26: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Above].
Line 25: Fenced code blocks should be surrounded by blank lines.

/home/work/pulumi/registry/themes/default/content/registry/packages/dynatrace/installation-configuration.md:
Line 28: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/eks/_index.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/eks/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/eventstorecloud/_index.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/exoscale/installation-configuration.md:
Line 18: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/flux/installation-configuration.md:
Line File Header: Meta description is too short. Must be at least 50 characters.

/home/work/pulumi/registry/themes/default/content/registry/packages/gandi/installation-configuration.md:
Line 26: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Above].
Line 25: Fenced code blocks should be surrounded by blank lines.

/home/work/pulumi/registry/themes/default/content/registry/packages/gcp/_index.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 11: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 13: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 16: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 121: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/gcp/installation-configuration.md:
Line 27: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 54: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 61: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/gcp-global-cloudrun/installation-configuration.md:
Line 20: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/google-cloud-static-website/_index.md:
Line 63: Multiple consecutive blank lines [Expected: 1; Actual: 2].
Line 66: Files should end with a single newline character.

/home/work/pulumi/registry/themes/default/content/registry/packages/google-cloud-static-website/installation-configuration.md:
Line 38: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Below].
Line 39: Fenced code blocks should be surrounded by blank lines.

/home/work/pulumi/registry/themes/default/content/registry/packages/google-native/_index.md:
Line 10: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 15: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/google-native/installation-configuration.md:
Line 22: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 28: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/grafana/installation-configuration.md:
Line 28: Trailing spaces [Expected: 0 or 2; Actual: 1].
Line 65: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/harbor/_index.md:
Line 125: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/harbor/installation-configuration.md:
Line 62: Unordered list style [Expected: asterisk; Actual: dash].
Line 63: Unordered list style [Expected: asterisk; Actual: dash].
Line 64: Unordered list style [Expected: asterisk; Actual: dash].
Line 65: Unordered list style [Expected: asterisk; Actual: dash].
Line 66: Unordered list style [Expected: asterisk; Actual: dash].

/home/work/pulumi/registry/themes/default/content/registry/packages/harness/_index.md:
Line 25: Fenced code blocks should be surrounded by blank lines.

/home/work/pulumi/registry/themes/default/content/registry/packages/harness/installation-configuration.md:
Line 70: Multiple consecutive blank lines [Expected: 1; Actual: 2].
Line 68: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/hcp/installation-configuration.md:
Line 56: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/koyeb/_index.md:
Line 150: Files should end with a single newline character.

/home/work/pulumi/registry/themes/default/content/registry/packages/koyeb/installation-configuration.md:
Line File Header: Meta description is too short. Must be at least 50 characters.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes/_index.md:
Line 11: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 40: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 81: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 91: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 97: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 113: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes/faq.md:
Line 19: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 68: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 163: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes/installation-configuration.md:
Line 73: Trailing spaces [Expected: 0 or 2; Actual: 1].
Line 8: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 46: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 63: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 65: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes/introduction.md:
Line 56: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes-cert-manager/installation-configuration.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes-coredns/installation-configuration.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/kubernetes-ingress-nginx/installation-configuration.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/launchdarkly/installation-configuration.md:
Line 65: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/metabase/_index.md:
Line 95: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 147: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 159: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/metabase/installation-configuration.md:
Line 9: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/nomad/installation-configuration.md:
Line 7: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/openstack/installation-configuration.md:
Line 8: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/proxmoxve/installation-configuration.md:
Line File Header: Page title exceeds 60 characters.

/home/work/pulumi/registry/themes/default/content/registry/packages/pulumiservice/installation-configuration.md:
Line 61: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/purrl/installation-configuration.md:
Line File Header: Meta description is too short. Must be at least 50 characters.

/home/work/pulumi/registry/themes/default/content/registry/packages/scaleway/installation-configuration.md:
Line 65: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/sentry/installation-configuration.md:
Line 18: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/symbiosis/_index.md:
Line 23: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/symbiosis/installation-configuration.md:
Line 26: Unordered list style [Expected: asterisk; Actual: dash].
Line 27: Unordered list style [Expected: asterisk; Actual: dash].
Line 29: Multiple consecutive blank lines [Expected: 1; Actual: 2].

/home/work/pulumi/registry/themes/default/content/registry/packages/synced-folder/_index.md:
Line 339: Headings should be surrounded by blank lines [Expected: 1; Actual: 0; Above].

/home/work/pulumi/registry/themes/default/content/registry/packages/synced-folder/installation-configuration.md:
Line 19: Multiple consecutive blank lines [Expected: 1; Actual: 2].
Line 56: Multiple consecutive blank lines [Expected: 1; Actual: 2].
Line 57: Multiple consecutive blank lines [Expected: 1; Actual: 3].
Line 52: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 54: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/tls-self-signed-cert/_index.md:
Line 104: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.
Line 114: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

/home/work/pulumi/registry/themes/default/content/registry/packages/unifi/_index.md:
Line 23: Trailing spaces [Expected: 0 or 2; Actual: 1].

/home/work/pulumi/registry/themes/default/content/registry/packages/unifi/installation-configuration.md:
Line 24: Trailing spaces [Expected: 0 or 2; Actual: 1].
Line File Header: Meta description is too short. Must be at least 50 characters.

/home/work/pulumi/registry/themes/default/content/registry/packages/upstash/installation-configuration.md:
Line 62: Using relrefs in registry are no longer supported. Use standard [Markdown](/links) instead.

    
```

susanev commented:
ah @sean1588 sry i didnt give you all the context you needed. we turned off linting because it was failing on all the community packages and we didn't want to put that burden onto our community.

but we do want to lint for things that will break the docs build, like relrefs. i think we can just start with relrefs for now, unless you know of some other things that can break the build?

sean1588 commented:
ahh no worries @susanev! I'll take a quick glance through some failed docs builds and see if I find any other things that stick out that might be breaking it and if not just isolate this to just the relrefs for now.

cnunciato commented:
I wonder if this sort of linting belongs here rather than at the upstream. For example, a failure here is generally not fixable here; things like docs errors and the like need to be fixed at the upstream -- and upstreams (presumably) don't have an easy way of knowing whether their PRs will fail once they get here. Maybe an action that does the validation we need would be a better solution? Thoughts?

sean1588 commented:
Ok I think I see what you are saying... By upstream, I am assuming you are referring to the provider/component repos, and some of those errors would get trapped here when the content gets synced? I think there is some content that is manually added to this repo that the relref linting might be useful for. As far as the upstream provider repos are concerned, I guess the only thing I can think of to prevent that is to have the logic wrapped up in an action like you say and then we can get the action added those repos (at least the ones we manage) and also get it added it to the provider and component boilerplates.

cnunciato commented:
>have the logic wrapped up in an action like you say and then we can get the action added those repos

That's what I had in mind, yep -- and I guess ideally, that same action could be used by anyone building community packages, not just us.

susanev commented:
i dont mind exploring that direction, but still would like to get this in before that, unless anyone has concerns. i can manage communicating with community folks.

sean1588 commented on themes/default/content/registry/packages/aws/_index.md:
```diff
@@ -6,14 +6,14 @@ layout: overview
 The Amazon Web Services (AWS) provider for Pulumi can provision many of the cloud resources available in [AWS](https://aws.amazon.com/). It uses the AWS SDK to manage and provision resources.
 
-The AWS provider must be configured with credentials to deploy and update resources in AWS; see [Installation & Configuration]({{<relref "./installation-configuration">}}) for instructions.
+The AWS provider must be configured with credentials to deploy and update resources in AWS; see [Installation & Configuration](https://pulumi.com./installation-configuration) for instructions.
```
ahh I need to handle this case where there is a "." prepending the relref

sean1588 commented:
ok, I updated the lint script to have a fix function so we can auto-fix any of the relref failures. I added a Makefile target, `make lint-fix` which will fix up the failures. I then ran the new fix function and had it fix up all the outstanding relref failures. So now everything is passing. So this PR has the new relref fix funtionality as well as the fixed files. This is not hooked up to the pipeline yet. I'll make those pipeline adjustments once we get this in and also have it run the fixer when changes come in from upstream.

sean1588 commented on themes/default/content/registry/packages/awsx/installation-configuration.md:
```diff
@@ -5,7 +5,7 @@ layout: installation
 
 <!-- markdownlint-disable url -->
-[Pulumi AWSx provider]: {{< relref "./" >}}
+[Pulumi AWSx provider]: ./
```
should we just get rid of these sections altogether now?
